armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py

```diff
@@ -0,0 +1,598 @@
+                                .type(torch.uint8)
+                                .to(self.device)
+                            )
+                        y_preprocessed_tensor.append(y_preprocessed_t)
+                    y_preprocessed = y_preprocessed_tensor
+
+                transform = torchvision.transforms.Compose(
+                    [torchvision.transforms.ToTensor()]
+                )
+                image_tensor_list_grad = []
+
+                for i in range(x_preprocessed.shape[0]):
+                    if self.clip_values is not None:
+                        x_grad = transform(x_preprocessed[i] / self.clip_values[1]).to(
+                            self.device
+                        )
+                    else:
+                        x_grad = transform(x_preprocessed[i]).to(self.device)
+                    x_grad.requires_grad = True
+                    image_tensor_list_grad.append(x_grad)
+
+                inputs_t = image_tensor_list_grad
+
+            else:
+                raise NotImplementedError(
+                    "Combination of inputs and preprocessing not supported."
+                )
+
+            if isinstance(y_preprocessed, np.ndarray):
+                labels_t = torch.from_numpy(y_preprocessed).to(self.device)  # type: ignore
+            else:
+                labels_t = y_preprocessed  # type: ignore
+
+            output = self._model(inputs_t, labels_t)
+
+            return output, inputs_t, image_tensor_list_grad
+
+    def _train_step(
+        self,
+        images: "torch.Tensor",
+        target: "torch.Tensor",
+        mask: Optional["torch.Tensor"] = None,
+    ) -> "torch.Tensor":
+        import torch  # lgtm [py/repeated-import]
+
+        self.estimator.model.zero_grad()
+        loss = self._loss(images, target, mask)
+        loss.backward(retain_graph=True)
+
+        if self._optimizer_string == "pgd":
+            patch_grads = self._patch.grad
+            if patch_grads.shape[0] == 6:
+                patch_grads[3:, :, :] = torch.mean(
+                    patch_grads[3:, :, :], dim=0, keepdim=True
+                )
+
+            gradients = patch_grads.sign() * self.learning_rate
+            if patch_grads.shape[0] == 6:
+                gradients[3:, :, :] = (
+                    gradients[3:, :, :] / self.learning_rate * self.learning_rate_depth
+                )
+
+            with torch.no_grad():
+                self._patch[:] = torch.clamp(
+                    self._patch + gradients,
+                    min=self.estimator.clip_values[0],
+                    max=self.estimator.clip_values[1],
+                )
+
+                if self._patch.shape[0] == 6:
+                    self._patch[3:, :, :] = torch.clamp(
+                        self._patch[3:, :, :], min=self.min_depth, max=self.max_depth
+                    )
+        else:
+            raise ValueError(
+                "Adam optimizer for CARLA Adversarial Patch not supported."
```

Comment:
is there meant to be modified code below the `ValueError` that won't be reached? If the only possible values for `self._optimizer_string` are `"pgd"` and `"Adam"`, then it seems like the following would make the most sense:

```
if "pgd":
    ...
elif "Adam":
    raise the ValueError not supported for CARLA
else:
    raise a ValueError for unrecognized _optimizer_string
```

Reply:
Initially, when I used `self.learning_rate` for updating both RGB and depth, the Adam optimizer can be used. Later when I decided that the depth channel needed a separate learning rate, i.e., `self.learning_rate_depth`, the Adam optimizer could no longer accommodate two different learning rates. That's why I added the `raise ValueError()` but forgot to delete everything after it.